src/main/kotlin/com/autonomousapps/internal/parse/KotlinBuildScriptDependenciesRewriter.kt

```diff
@@ -0,0 +1,176 @@
+import cash.grammar.kotlindsl.utils.CollectingErrorListener
+import cash.grammar.kotlindsl.utils.Context.leafRule
+import cash.grammar.kotlindsl.utils.DependencyExtractor
+import cash.grammar.kotlindsl.utils.Whitespace
+import com.autonomousapps.exception.BuildScriptParseException
+import com.autonomousapps.internal.advice.AdvicePrinter
+import com.autonomousapps.internal.utils.filterToOrderedSet
+import com.autonomousapps.internal.utils.filterToSet
+import com.autonomousapps.internal.utils.ifNotEmpty
+import com.autonomousapps.model.Advice
+import com.squareup.cash.grammar.KotlinParser.NamedBlockContext
+import com.squareup.cash.grammar.KotlinParser.PostfixUnaryExpressionContext
+import com.squareup.cash.grammar.KotlinParser.ScriptContext
+import com.squareup.cash.grammar.KotlinParserBaseListener
+import org.antlr.v4.runtime.CharStream
+import org.antlr.v4.runtime.CommonTokenStream
+import java.nio.file.Path
+
+/**
+ * Rewrites a Kotlin build script based on provided advice, which includes:
+ *
+ * - When entering the `dependencies` block, processing each dependency declaration to either remove or update it according to the advice provided.
+ * - Upon exiting the `dependencies` block, adding any new dependencies.
+ * - At the end of the script, adding a `dependencies` block if new dependencies are present and the block is currently missing.
+ *
+ * Note: Buildscript dependencies are ignored.
+ */
+internal class KotlinBuildScriptDependenciesRewriter(
+  private val input: CharStream,
+  private val tokens: CommonTokenStream,
+  private val errorListener: CollectingErrorListener,
+  private val advice: Set<Advice>,
+  private val printer: AdvicePrinter,
+  /** Reverse map from custom representation to standard. */
+  private val reversedDependencyMap: (String) -> String,
+) : BuildScriptDependenciesRewriter, KotlinParserBaseListener() {
+  private val rewriter = Rewriter(tokens)
+  private val indent = Whitespace.computeIndent(tokens, input)
+
+  private val dependencyExtractor = DependencyExtractor(
+    input = input,
+    tokens = tokens,
+    indent = indent,
+  )
+
+  private var hasDependenciesBlock = false
+  private var inBuildscriptBlock = false
+
+  /**
+   * Returns build script with advice applied.
+   *
+   * Throws [BuildScriptParseException] if the script has some idiosyncrasy that impairs parsing.
+   *
+   */
+  @Throws(BuildScriptParseException::class)
+  override fun rewritten(): String {
+    errorListener.getErrorMessages().ifNotEmpty {
+      throw BuildScriptParseException.withErrors(it)
+    }
+
+    return rewriter.text
+  }
+
+  override fun enterNamedBlock(ctx: NamedBlockContext) {
+    dependencyExtractor.onEnterBlock()
+
+    if (ctx.isBuildscript) {
+      inBuildscriptBlock = true
+    }
+
+    if (ctx.isDependencies) {
+      handleDependencies(ctx)
+    }
+  }
+
+  override fun exitNamedBlock(ctx: NamedBlockContext) {
+    dependencyExtractor.onExitBlock()
```

Comment:
I think the (implied) contract of this function is that it should be at the bottom of the `exitNamedBlock` function.

Reply:
Done in https://github.com/autonomousapps/dependency-analysis-gradle-plugin/pull/1299/commits/d36ba8a9cf5cf6d03ba51863f32512ff38002ced#diff-b2a94f740ad8ff6516bf944965d212e085314087d0edc9cc852710ad1ef35c21R94-R95